### PR TITLE
allow custom user renovate configs, in config maps,

### PR DESCRIPTION
**Pull request by rcerven**
there can be one global one for all components, and also custom config for specific components

we are also now creating renovate config in json format

[STONEBLD-2916](https://issues.redhat.com//browse/STONEBLD-2916)

### Checklist:
 - PR has reference to the issue(s) it resolves
 - Write / update unit tests
 - Write / update integration (envtest) tests
 - Ensure there is an issue for e2e tests if needed
 - Ensure `make test` passes
 - Ensure test coverage hasn't decreased
 - Test code changes manually
 - Update readme and documentation
 - Write PR description that highlights overall changes and add usage examples if applicable

### CODE REVIEW

File: controllers/renovate_util.go [hunk 1]
```diff
@@ -380,21 +376,73 @@ func (u ComponentDependenciesUpdater) CreateRenovaterPipeline(ctx context.Contex
...
+		if userConfigMap.Name == GlobalRenovateConfigName {
+			globalConfigString, globalConfigType = getConfigAndTypeFromConfigMap(userConfigMap)
+			break
+		}
+	}
+
 	for _, target := range targets {
 		randomStr1 := RandomString(5)
 		randomStr2 := RandomString(10)
 		randomStr3 := RandomString(10)
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/renovate_util.go:396]: well those are really just random strings, so not changing that

File: controllers/renovate_util.go [hunk 2]
```diff
@@ -380,21 +376,73 @@ func (u ComponentDependenciesUpdater) CreateRenovaterPipeline(ctx context.Contex
...
+			if componentConfigString != "" {
+				configString = componentConfigString
+				configType = componentConfigType
+				log.Info("will use custom renovate config for component", "name", componentConfigName, "type", configType)
+			} else {
+				configString = globalConfigString
+				configType = globalConfigType
+				log.Info("will use custom global renovate config", "name", GlobalRenovateConfigName, "type", configType)
+			}
+
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/renovate_util.go:422]: ok

File: controllers/renovate_util.go [hunk 3]
```diff
@@ -246,116 +290,68 @@ func (u ComponentDependenciesUpdater) GetUpdateTargetsGithubApp(ctx context.Cont
...
-	}
+	var matchStrings []string
+	var registryAliases = make(map[string]string)
+	var customManagers []CustomManager
+	var packageRules []PackageRule
+	var matchPackageNames []string
+	matchStrings = append(matchStrings, buildResult.BuiltImageRepository+"(:.*)?@(?<currentDigest>sha256:[a-f0-9]+)")
+	matchPackageNames = append(matchPackageNames, buildResult.BuiltImageRepository)
+
+	for _, drepositiry := range buildResult.DistributionRepositories {
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/renovate_util.go:307]: ok

File: controllers/component_dependency_update_controller_test.go [hunk 4]
```diff
@@ -245,6 +240,8 @@ var _ = Describe("Component nudge controller", func() {
 
 			renovatePipelines := getRenovatePipelineRunList()
 			Expect(len(renovatePipelines)).Should(Equal(1))
+			renovateCommand := strings.Join(renovatePipelines[0].Spec.PipelineSpec.Tasks[0].TaskSpec.Steps[0].Command, ";")
+			Expect(strings.Contains(renovateCommand, `'username':'image_repo_username'`)).Should(BeTrue())
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/component_dependency_update_controller_test.go:244]: well tbh this is used mostly in the code and seems quite readable

File: controllers/component_dependency_update_controller_test.go [hunk 5]
```diff
@@ -416,19 +403,186 @@ var _ = Describe("Component nudge controller", func() {
...
+		It("Test build performs nudge on success, only component renovate config provided", func() {
+			customConfigName := fmt.Sprintf("nudging-renovate-config-%s", Operator1)
+			customConfigMapName1 := types.NamespacedName{Namespace: UserNamespace, Name: customConfigName}
+			customConfigString := `{"username":"componentconfiguserjs"}`
+			customConfigMapData := map[string]string{ConfigKeyJs: customConfigString}
+			createCustomRenovateConfigMap(customConfigMapName1, customConfigMapData)
+			customConfigName = fmt.Sprintf("nudging-renovate-config-%s", Operator2)
+			customConfigMapName2 := types.NamespacedName{Namespace: UserNamespace, Name: customConfigName}
+			createCustomRenovateConfigMap(customConfigMapName2, customConfigMapData)
+			customConfigType := "js"
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/component_dependency_update_controller_test.go:465]: ok

**Comment by rcerven** [comment]
related docs PR: https://github.com/konflux-ci/docs/pull/171

**Review by mmorhun — APPROVED** [review]

File: controllers/renovate_util.go [hunk 6]
```diff
@@ -380,21 +376,70 @@ func (u ComponentDependenciesUpdater) CreateRenovaterPipeline(ctx context.Contex
 	secretTokens := map[string]string{}
 	configmaps := map[string]string{}
 	renovateCmds := []string{}
+	globalConfigString := ""
+	globalConfigType := ""
+
+	allUserConfigMaps := &corev1.ConfigMapList{}
+	if err := u.Client.List(ctx, allUserConfigMaps, client.InNamespace(namespace)); err != nil {
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/renovate_util.go:383]: I wouldn't complicate it more for users tbh

sure finding could be filtered, but on other hand with missing label users would complain that it doesn't work
so better less things for them to set

File: controllers/suite_util_test.go [hunk 7]
```diff
@@ -635,3 +646,19 @@ func deletePipelineServiceAccount(namespace string) {
 	pipelineServiceAccountkey := types.NamespacedName{Name: buildPipelineServiceAccountName, Namespace: namespace}
 	deleteServiceAccount((pipelineServiceAccountkey))
 }
+
+func deleteConfigMap(configMapKey types.NamespacedName) {
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/suite_util_test.go]: moved

**Review by tnevrlka — APPROVED** [review]

**Pull request merged**